pkg/localstorev2/recover.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storer
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/localstorev2/internal/chunkstore"
+	"github.com/ethersphere/bee/pkg/sharky"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	sharkyDirtyFileName = ".DIRTY"
+)
+
+func sharkyRecovery(ctx context.Context, sharkyBasePath string, store storage.Store, opts *Options) (closerFn, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	logger := opts.Logger.WithName(loggerName).Register()
+	dirtyFilePath := filepath.Join(sharkyBasePath, sharkyDirtyFileName)
+
+	closer := func() error { return os.Remove(dirtyFilePath) }
+
+	if _, err := os.Stat(dirtyFilePath); errors.Is(err, fs.ErrNotExist) {
+		return closer, os.WriteFile(dirtyFilePath, []byte{}, 0644)
+	}
+
+	logger.Info("localstore sharky .DIRTY file exists: starting recovery due to previous dirty exit")
+	defer func(t time.Time) {
+		logger.Info("localstore sharky recovery finished", "time", time.Since(t))
+	}(time.Now())
+
+	sharkyRecover, err := sharky.NewRecovery(sharkyBasePath, sharkyNoOfShards, swarm.SocMaxChunkSize)
+	if err != nil {
+		return closer, err
+	}
+
+	defer func() {
+		if err := sharkyRecover.Close(); err != nil {
+			logger.Error(err, "failed closing sharky recovery")
+		}
+	}()
+
+	locationResultC := make(chan chunkstore.LocationResult, 128)
```

Comment:
I dont see why we need this buffered channel. There is only 1 reader, so operations will only happen sequentially. See the other comments about naked writes. This entire operation will be controlled by context cancellation. So we can handle it immediately without using the buffered channel.

Reply:
Buffer was added in order to decouple goroutine in `chunkstore.IterateLocations` from goroutine running `addLocations`. 

It doesn't matter how many readers are there, but if reader is slower it will unnecessarily block writer which will impair performance.

Reply:
So buffer of 128 does not guarantee unblocking anything. If there are millions of locations things could still block. So these type of magic numbers need to be avoided unless we know exactly why and how many we need.

Reply:
Correct, for every number that is picked as capacity for buffer there could be case were this capacity is not going to be sufficient. 

However, adding small capacity for channel can still significantly improve perfs because reader and writer are going to decouple it's executions.

Reply:
But this is not required. There is no evidence this will improve anything. Having a buffer can lead to producer overwhelming the consumer and taking CPU cycles from it. This might lead to an overall degradation in performance. When there are two routines interacting it is better to ensure they peform according to their speed and block the other. This is the most simple design and to change this there has to be some data to make any claims.

Reply:
I forgot to mention that buffer has been removed from this channel.

We can continue this discussion in dedicated call on this topic.